network/crl.go

```diff
@@ -0,0 +1,380 @@
+/*
+Copyright 2020, Cossack Labs Limited
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	"github.com/golang/groupcache/lru"
+)
+
+// --tls_crl_from_cert=<use|ignore>
+const (
+	// If certificate contains CRL distribution point(s), use them
+	crlFromCertUseStr = "use"
+	// Ignore CRL distribution points listed in certificate
+	crlFromCertIgnoreStr = "ignore"
+)
+
+const (
+	crlFromCertUse int = iota
+	crlFromCertIgnore
+)
+
+// CRLConfig contains configuration related to certificate validation using CRL
+type CRLConfig struct {
+	uri      string
+	fromCert int // crlFromCert*
+}
+
+// NewCRLConfig creates new CRLConfig
+func NewCRLConfig(uri, fromCert string) (*CRLConfig, error) {
+	_, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	var fromCertVal int
+	switch fromCert {
+	case crlFromCertUseStr:
+		fromCertVal = crlFromCertUse
+	case crlFromCertIgnoreStr:
+		fromCertVal = crlFromCertIgnore
+	default:
+		return nil, errors.New("Invalid `tls_crl_from_cert` value '" + fromCert + "', should be one of 'use', 'ignore'")
+	}
+
+	if uri != "" {
+		// Since this is CRL configuration alone, we don't have access to cache yet;
+		// so let's just download the CRL and forget about it
+		_, err = DefaultCRLClient.Fetch(DefaultCRLClient{}, uri)
+		if err != nil {
+			log.WithError(err).Warnf("CRL: Cannot fetch configured URI '%s'", uri)
```

Comment:
For this I've put a TODO:

	// TODO return error after issues with failing tests are fixed;
	//      CRL HTTP server is starting, connection is checked, then Acra is starting
	//      but somehow checking connection *here* fails thus failing the tests;
	//      everything else seems working since real requests to configured server
	//      are successful (when CRL verification is performed)
	// return nil, fmt.Errorf("CRL: Cannot fetch configured URI '%s'", uri)